File: Measurement Studio/Examples/DotNET/v4.5/Applications/Application/UISnippets/vb/Legend.vb
Imports NationalInstruments.UI.WindowsForms

Class SnipsLegend
    Inherits SnipsControlBase
    Private legend As Legend

    Public Sub New(ByVal Legend__1 As Legend)
        MyBase.New(Legend__1)
        legend = Legend__1
    End Sub

#Region "code snippets for NationalInstruments.UI.WindowsForms.Legend"

    ''' <summary>
    ''' Returns a LegendHitTestInfo that specifies where on the control the given
    ''' point is located.  It is implemented in the Legend class. To run this method,
    ''' you must first click the run snippet button, and then click somewhere inside 
    ''' the legend area. 
    ''' </summary>
    ''' <signature>HitTest(Integer, Integer)</signature>
    ''' <OtherMethods>
    ''' Legend.GetItemAt(Integer, Integer)
    ''' </OtherMethods>
    ''' <ExampleMethod />
    <EventBased("MouseDown")> _
    Public Sub Legend_HitTest_int_int(ByVal sender As Object, ByVal e As MouseEventArgs)
        ' The following example demonstrates using the HitTest method to determine
        ' where a user clicked on a Legend object.
        Dim hitTestRegion As LegendHitTestInfo
        Dim randomColor As Color = Color.FromArgb(RandNumberGenerator.[Next](255), RandNumberGenerator.[Next](255), RandNumberGenerator.[Next](255))

        hitTestRegion = legend.HitTest(e.X, e.Y)
        Select Case hitTestRegion
            Case LegendHitTestInfo.HorizontalScrollBar
                Debug.WriteLine("horizontal scrollbar selected")
                Exit Select
            Case LegendHitTestInfo.Item
                Dim item As LegendItem = legend.GetItemAt(e.X, e.Y)
                If TypeOf item.Source Is Plot Then
                    Dim plot As Plot = DirectCast(item.Source, Plot)
                    plot.LineColor = randomColor
                    Debug.WriteLine("Item selected was a Plot")
                ElseIf TypeOf item.Source Is DigitalPlot Then
                    Dim plot As DigitalPlot = DirectCast(item.Source, DigitalPlot)
                    plot.LineColor = randomColor
                    Debug.WriteLine("Item selected was a Digital Plot")
                End If
                Exit Select
            Case LegendHitTestInfo.Text
                Debug.WriteLine("Text area of legend selected")
                Exit Select
            Case LegendHitTestInfo.VerticalScrollBar
                Debug.WriteLine("vertical scrollbar selected")
                Exit Select
            Case LegendHitTestInfo.None
                Debug.WriteLine("Unknown legend area selected")
                Exit Select
        End Select
    End Sub

    ''' <summary>
    ''' Retrieves the size of a rectangular area Integero which the control can 
    ''' be fitted. This member overrides Control.GetPreferredSize and is 
    ''' implemented in the Legend class.  
    ''' </summary>
    ''' <signature>Legend.GetPreferredSize(Size)</signature>
    ''' <ExampleMethod />
    Public Sub GetPreferredSize_Size()
        ' The following example demonstrates getting the preferred size of a Legend object.
        Dim newSize As Size = legend.GetPreferredSize(Size.Empty)

        Debug.WriteLine(String.Format("The legend's original size is {0}", legend.Size.ToString()))
        Debug.WriteLine(String.Format("The legend's proposed size is {0}", newSize.ToString()))
    End Sub

#End Region

#Region "helper methods for the SnipsLegend class"

    Public Overrides Function ToString() As String
        Return legend.ToString()
    End Function

    ''' <summary>
    ''' Updates the legend to display the items in the SnipsLegendItem list
    ''' </summary>
    ''' <param name="items">Items to be displayed in the legend.</param>
    Public Sub SetItems(ByVal items As List(Of SnipsLegendItem))
        Dim legendItem As LegendItem

        legend.Items.Clear()

        If items IsNot Nothing Then
            For Each item As SnipsLegendItem In items
                If item.IsVisible Then
                    legendItem = New LegendItem(item, item.ItemLabel)
                    legend.Items.Add(legendItem)
                End If
            Next
        End If
    End Sub
#End Region
End Class

#Region "helper methods for the SnipsLegend class"
''' <summary>
''' A helper class for snips controls that need to be added to a legend
''' </summary>
Public Class SnipsLegendItem
    Implements ILegendItemSource
    Private item As ILegendItemSource
    Private label As String
    Private visible As Boolean

    ''' <summary>
    ''' Public constructor for the Snips Legend Item
    ''' </summary>
    ''' <param name="Item">The item to be added to the legend</param>
    ''' <param name="ItemLabel">The label to be displayed on the legend</param>
    ''' <param name="Visible">Whether or not the item is visible</param>
    Public Sub New(ByVal item As ILegendItemSource, ByVal itemLabel As String, ByVal visible As Boolean)
        Me.item = item
        Me.label = itemLabel
        Me.visible = visible
    End Sub

    ''' <summary>
    ''' Draws the symbol of a legend item
    ''' </summary>
    ''' <param name="args">A ComponentDrawArgs that contains the graphics 
    ''' surface to draw the legend item on and the bounds in which to 
    ''' draw the legend item.</param>
    Public Sub DrawLegendItem(ByVal args As ComponentDrawArgs) Implements ILegendItemSource.DrawLegendItem
        item.DrawLegendItem(args)
    End Sub

    ''' <summary>
    ''' Event signaling when the legend item has been disposed
    ''' </summary>
    Public Event Disposed As EventHandler Implements ILegendItemSource.Disposed
    Protected Overridable Sub OnDisposed(ByVal e As EventArgs)
        RaiseEvent Disposed(Me, e)
    End Sub

    ''' <summary>
    ''' Event signaling when the legend item changes
    ''' </summary>
    Public Event LegendItemChanged As EventHandler Implements ILegendItemSource.LegendItemChanged
    Protected Overridable Sub OnLegendItemChanged(ByVal e As EventArgs)
        RaiseEvent LegendItemChanged(Me, e)
    End Sub

    ''' <summary>
    ''' The label to be displayed for the legend item
    ''' </summary>
    Public Property ItemLabel() As String
        Get
            Return label
        End Get
        Set(ByVal value As String)
            label = value
        End Set
    End Property

    ''' <summary>
    ''' Whether or not the item is visible
    ''' </summary>
    Public Property IsVisible() As Boolean
        Get
            Return visible
        End Get
        Set(ByVal value As Boolean)
            visible = value
        End Set
    End Property
End Class
#End Region

File: Measurement Studio/Examples/DotNET/v4.5/Applications/Application/UISnippets/cs/Legend.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Drawing;
using System.Windows.Forms;
using NationalInstruments.UI;
using NationalInstruments.UI.WindowsForms;

namespace NationalInstruments.Examples.Snippets
{
    class SnipsLegend : SnipsControlBase
    {
        private Legend legend;

        public SnipsLegend(Legend Legend)
            : base (Legend)
        {
            legend = Legend;            
        }

        #region code snippets for NationalInstruments.UI.WindowsForms.Legend

        /// <summary>
        /// Returns a LegendHitTestInfo that specifies where on the control the given
        /// point is located.  It is implemented in the Legend class. To run this method,
        /// you must first click the run snippet button, and then click somewhere inside 
        /// the legend area. 
        /// </summary>
        /// <signature>HitTest(int, int)</signature>
        /// <OtherMethods>
        /// Legend.GetItemAt(int, int)
        /// </OtherMethods>
        /// <ExampleMethod />
        [EventBased("MouseDown")]
        public void Legend_HitTest_int_int(object sender, MouseEventArgs e)
        {
            // The following example demonstrates using the HitTest method to determine
            // where a user clicked on a Legend object.
            LegendHitTestInfo hitTestRegion;
            Color randomColor = Color.FromArgb(RandNumberGenerator.Next(255),
                                                RandNumberGenerator.Next(255),
                                                RandNumberGenerator.Next(255));

            hitTestRegion = legend.HitTest(e.X, e.Y);
            switch (hitTestRegion)
            {
                case LegendHitTestInfo.HorizontalScrollBar:
                    Debug.WriteLine("horizontal scrollbar selected");                    
                    break;
                case LegendHitTestInfo.Item:
                    LegendItem item = legend.GetItemAt(e.X, e.Y);
                    if (item.Source is Plot)
                    {
                        Plot plot = (Plot)item.Source;
                        plot.LineColor = randomColor;
                        Debug.WriteLine("Item selected was a Plot");
                    }
                    else if (item.Source is DigitalPlot)
                    {
                        DigitalPlot plot = (DigitalPlot)item.Source;
                        plot.LineColor = randomColor;
                        Debug.WriteLine("Item selected was a Digital Plot");
                    }
                    break;
                case LegendHitTestInfo.Text:
                    Debug.WriteLine("Text area of legend selected");
                    break;
                case LegendHitTestInfo.VerticalScrollBar:
                    Debug.WriteLine("vertical scrollbar selected");
                    break;
                case LegendHitTestInfo.None:
                    Debug.WriteLine("Unknown legend area selected");
                    break;
            }
        }

        /// <summary>
        /// Retrieves the size of a rectangular area into which the control can 
        /// be fitted. This member overrides Control.GetPreferredSize and is 
        /// implemented in the Legend class.  
        /// </summary>
        /// <signature>Legend.GetPreferredSize(Size)</signature>
        /// <ExampleMethod />
        public void GetPreferredSize_Size()
        {
            // The following example demonstrates getting the preferred size of a Legend object.
            Size newSize = legend.GetPreferredSize(Size.Empty);            

            Debug.WriteLine(string.Format("The legend's original size is {0}", legend.Size.ToString()));
            Debug.WriteLine(string.Format("The legend's proposed size is {0}", newSize.ToString()));
        }

        #endregion

        #region helper methods and classes for the SnipsLegend class

        public override string ToString()
        {
            return legend.ToString();
        }

        /// <summary>
        /// Updates the legend to display the items in the SnipsLegendItem list
        /// </summary>
        /// <param name="items">Items to be displayed in the legend.</param>
        public void SetItems(List<SnipsLegendItem> items)
        {
            LegendItem legendItem;            

            legend.Items.Clear();

            if (items != null)
            {
                foreach (SnipsLegendItem item in items)
                {
                    if (item.IsVisible)
                    {
                        legendItem = new LegendItem(item, item.ItemLabel);
                        legend.Items.Add(legendItem);
                    }
                }
            }
        }
    }

    /// <summary>
    /// A helper class for snips controls that need to be added to a legend
    /// </summary>
    public class SnipsLegendItem : ILegendItemSource
    {
        private ILegendItemSource item;
        private string itemLabel;
        private bool visible;

        /// <summary>
        /// Public constructor for the Snips Legend Item
        /// </summary>
        /// <param name="item">The item to be added to the legend</param>
        /// <param name="itemLabel">The label to be displayed on the legend</param>
        /// <param name="visible">Whether or not the item is visible</param>
        public SnipsLegendItem(ILegendItemSource item, string itemLabel, bool visible)
        {
            this.item = item;
            this.itemLabel = itemLabel;
            this.visible = visible;
        }

        /// <summary>
        /// Draws the symbol of a legend item
        /// </summary>
        /// <param name="args">A ComponentDrawArgs that contains the graphics 
        /// surface to draw the legend item on and the bounds in which to 
        /// draw the legend item.</param>
        public void DrawLegendItem(ComponentDrawArgs args)
        {
            item.DrawLegendItem(args);
        }

        /// <summary>
        /// Event signaling when the legend item has been disposed
        /// </summary>
        public event EventHandler Disposed;
        /// <summary>
        /// Raises the Disposed event
        /// </summary>
        /// <param name="e">Contains event data</param>
        protected virtual void OnDisposed(EventArgs e)
        {
            if (Disposed != null)
                Disposed(this, e);
        }

        /// <summary>
        /// Event signaling when the legend item changes
        /// </summary>
        public event EventHandler LegendItemChanged;
        /// <summary>
        /// Raises the LegendItemChanged event
        /// </summary>
        /// <param name="e">Contains event data</param>
        protected virtual void OnLegendItemChanged(EventArgs e)
        {
            if (LegendItemChanged != null)
                LegendItemChanged(this, e);
        }

        /// <summary>
        /// The label to be displayed for the legend item
        /// </summary>
        public string ItemLabel
        {
            get { return itemLabel; }
            set { itemLabel = value; }
        }

        /// <summary>
        /// Whether or not the item is visible
        /// </summary>
        public bool IsVisible
        {
            get { return visible; }
            set { visible = value; }
        }
    }
    #endregion
}
